gateway/host_checker.go

```diff
@@ -131,42 +133,55 @@ func (h *HostUptimeChecker) HostReporter() {
 	for {
 		select {
 		case okHost := <-h.okChan:
-			// Clear host from unhealthylist if it exists
-			if h.unHealthyList[okHost.CheckURL] {
-				newVal := 1
-				if count, found := h.sampleCache.Get(okHost.CheckURL); found {
-					newVal = count.(int) - 1
-				}
-
-				if newVal <= 0 {
-					// Reset the count
-					h.sampleCache.Delete(okHost.CheckURL)
-					log.Warning("[HOST CHECKER] [HOST UP]: ", okHost.CheckURL)
-					h.upCallback(okHost)
-					delete(h.unHealthyList, okHost.CheckURL)
-				} else {
-					log.Warning("[HOST CHECKER] [HOST UP BUT NOT REACHED LIMIT]: ", okHost.CheckURL)
-					h.sampleCache.Set(okHost.CheckURL, newVal, cache.DefaultExpiration)
+			// check if the the host url is in the sample map
+			if hostSample, found := h.samples.Load(okHost.CheckURL); found {
+				sample := hostSample.(HostSample)
+				//if it reached the h.sampleTriggerLimit, we're going to start decreasing the count value
+				if sample.reachedLimit {
+					newCount := sample.count - 1
+
+					if newCount <= 0 {
+						//if the count-1 is equals to zero, it means that the host is fully up.
+
+						h.samples.Delete(okHost.CheckURL)
+						log.Warning("[HOST CHECKER] [HOST UP]: ", okHost.CheckURL)
+						h.upCallback(okHost)
+					} else {
+						//in another case, we are one step closer. We just update the count number
+						sample.count = newCount
+						log.Warning("[HOST CHECKER] [HOST UP BUT NOT REACHED LIMIT]: ", okHost.CheckURL)
+						h.samples.Store(okHost.CheckURL, sample)
+					}
 				}
 			}
 			go h.pingCallback(okHost)
 
 		case failedHost := <-h.errorChan:
-			newVal := 1
-			if count, found := h.sampleCache.Get(failedHost.CheckURL); found {
-				newVal = count.(int) + 1
+			sample := HostSample{
+				count: 1,
+			}
+
+			//If a host fails, we check if it has failed already
+			if hostSample, found := h.samples.Load(failedHost.CheckURL); found {
+				sample = hostSample.(HostSample)
+				// we add THIS failure to the count
+				sample.count = sample.count + 1
 			}
 
-			if newVal >= h.sampleTriggerLimit {
+			if sample.count == h.sampleTriggerLimit {
```

Comment:
Application logic has changed here. 
Previously it was `sample.count < h.sampleTriggerLimit` or to be exact `else` statement of `newVal >= h.sampleTriggerLimit`

Reply:
@buger  I thought it wasn't necessary but I forgot about the fallback call. Good catch! 
Updated.